File: app/database/models/server/channel.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {children, field, immutableRelation} from '@nozbe/watermelondb/decorators';
import Model, {type Associations} from '@nozbe/watermelondb/Model';

import {MM_TABLES} from '@constants/database';

import type {Query, Relation} from '@nozbe/watermelondb';
import type CategoryChannelModel from '@typings/database/models/servers/category_channel';
import type ChannelModelInterface from '@typings/database/models/servers/channel';
import type ChannelInfoModel from '@typings/database/models/servers/channel_info';
import type ChannelMembershipModel from '@typings/database/models/servers/channel_membership';
import type DraftModel from '@typings/database/models/servers/draft';
import type MyChannelModel from '@typings/database/models/servers/my_channel';
import type PostModel from '@typings/database/models/servers/post';
import type PostsInChannelModel from '@typings/database/models/servers/posts_in_channel';
import type TeamModel from '@typings/database/models/servers/team';
import type UserModel from '@typings/database/models/servers/user';

const {
    CATEGORY_CHANNEL,
    CHANNEL,
    CHANNEL_INFO,
    CHANNEL_MEMBERSHIP,
    DRAFT,
    MY_CHANNEL,
    POSTS_IN_CHANNEL,
    POST,
    TEAM,
    USER,
} = MM_TABLES.SERVER;

/**
 * The Channel model represents a channel in the Mattermost app.
 */
export default class ChannelModel extends Model implements ChannelModelInterface {
    /** table (name) : Channel */
    static table = CHANNEL;

    /** associations : Describes every relationship to this table. */
    static associations: Associations = {

        /** A CHANNEL can be associated with multiple CHANNEL_MEMBERSHIP (relationship is 1:N) */
        [CHANNEL_MEMBERSHIP]: {type: 'has_many', foreignKey: 'channel_id'},

        /** A CHANNEL can be associated with one CATEGORY_CHANNEL per team (relationship is 1:1)  */
        [CATEGORY_CHANNEL]: {type: 'has_many', foreignKey: 'channel_id'},

        /** A CHANNEL can be associated with multiple DRAFT (relationship is 1:N) */
        [DRAFT]: {type: 'has_many', foreignKey: 'channel_id'},

        /** A CHANNEL can be associated with multiple POSTS_IN_CHANNEL (relationship is 1:N) */
        [POSTS_IN_CHANNEL]: {type: 'has_many', foreignKey: 'channel_id'},

        /** A CHANNEL can contain multiple POST (relationship is 1:N) */
        [POST]: {type: 'has_many', foreignKey: 'channel_id'},

        /** A CHANNEL can be associated to one MY_CHANNEL (relationship is 1:1) */
        [MY_CHANNEL]: {type: 'has_many', foreignKey: 'id'},

        /** A TEAM can be associated to CHANNEL (relationship is 1:N) */
        [TEAM]: {type: 'belongs_to', key: 'team_id'},

        /** A USER can create multiple CHANNEL (relationship is 1:N) */
        [USER]: {type: 'belongs_to', key: 'creator_id'},

        /** A CHANNEL is associated with one CHANNEL_INFO**/
        [CHANNEL_INFO]: {type: 'has_many', foreignKey: 'id'},

    };

    /** create_at : The creation date for this channel */
    @field('create_at') createAt!: number;

    /** creator_id : The user who created this channel */
    @field('creator_id') creatorId!: string;

    /** update_at : The timestamp to when this channel was last updated on the server */
    @field('update_at') updateAt!: number;

    /** delete_at : The deletion/archived date of this channel */
    @field('delete_at') deleteAt!: number;

    /** display_name : The channel display name (e.g. Town Square ) */
    @field('display_name') displayName!: string;

    /** is_group_constrained : If a channel is  restricted to certain groups, this boolean will be true and only
     * members of that group have access to this team. Hence indicating that the members of this channel are
     * managed by groups.
     */
    @field('is_group_constrained') isGroupConstrained!: boolean;

    /** name : The name of the channel (e.g town-square) */
    @field('name') name!: string;

    /** shared: determines if it is a shared channel with another organization */
    @field('shared') shared!: boolean;

    /** team_id : The team to which this channel belongs.  It can be empty for direct/group message. */
    @field('team_id') teamId!: string;

    /** type : The type of the channel ( e.g. G: group messages, D: direct messages, P: private channel and O: public channel) */
    @field('type') type!: ChannelType;

    /** members : Users belonging to this channel */
    @children(CHANNEL_MEMBERSHIP) members!: Query<ChannelMembershipModel>;

    /** drafts : All drafts for this channel */
    @children(DRAFT) drafts!: Query<DraftModel>;

    /** posts : All posts made in that channel */
    @children(POST) posts!: Query<PostModel>;

    /** postsInChannel : a section of the posts for that channel bounded by a range */
    @children(POSTS_IN_CHANNEL) postsInChannel!: Query<PostsInChannelModel>;

    /** team : The TEAM to which this CHANNEL belongs */
    @immutableRelation(TEAM, 'team_id') team!: Relation<TeamModel>;

    /** creator : The USER who created this CHANNEL*/
    @immutableRelation(USER, 'creator_id') creator!: Relation<UserModel>;

    /** info : Query returning extra information about this channel from CHANNEL_INFO table */
    // @lazy info = this.collections.get<ChannelInfoModel>(CHANNEL_INFO).query(Q.on(CHANNEL, 'id', this.id));
    @immutableRelation(CHANNEL_INFO, 'id') info!: Relation<ChannelInfoModel>;

    /** membership : Query returning the membership data for the current user if it belongs to this channel */
    @immutableRelation(MY_CHANNEL, 'id') membership!: Relation<MyChannelModel>;

    /** categoryChannel : Query returning the membership data for the current user if it belongs to this channel */
    @immutableRelation(CATEGORY_CHANNEL, 'channel_id') categoryChannel!: Relation<CategoryChannelModel>;

    toApi = (): Channel => {
        return {
            id: this.id,
            create_at: this.createAt,
            update_at: this.updateAt,
            delete_at: this.deleteAt,
            team_id: this.teamId,
            type: this.type,
            display_name: this.displayName,
            name: this.name,
            header: '',
            purpose: '',
            last_post_at: 0,
            total_msg_count: 0,
            extra_update_at: 0,
            creator_id: this.creatorId,
            scheme_id: null,
            group_constrained: null,
            shared: this.shared,
        };
    };
}

File: ios/Gekidou/Sources/Gekidou/DataTypes/Channel.swift
import Foundation

public struct Channel: Codable {
    let id: String
    let createAt: Double
    let creatorId: String
    let deleteAt: Double
    var displayName: String = ""
    let extraUpdateAt: Double
    let groupConstrained: Bool
    let header: String
    let lastPostAt: Double
    let lastRootPostAt: Double
    let name: String
    let policyId: String
    let props: String
    let purpose: String
    let schemeId: String
    let shared: Bool
    let teamId: String
    let totalMsgCount: Int
    let totalMsgCountRoot: Int
    let type: String
    let updateAt: Double
    
    public enum ChannelKeys: String, CodingKey {
        case id
        case createAt = "create_at"
        case creatorId = "creator_id"
        case deleteAt = "delete_at"
        case displayName = "display_name"
        case extraUpdateAt = "extra_update_at"
        case groupConstrained = "group_constrained"
        case header
        case lastPostAt = "last_post_at"
        case lastRootPostAt = "last_root_post_at"
        case name
        case policyId = "policy_id"
        case props
        case purpose
        case schemeId = "scheme_id"
        case shared
        case teamId = "team_id"
        case totalMsgCount = "total_msg_count"
        case totalMsgCountRoot = "total_msg_count_root"
        case type
        case updateAt = "update_at"
    }
    
    public init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: ChannelKeys.self)
        id  = try values.decode(String.self, forKey: .id)
        creatorId = values.decodeIfPresent(forKey: .creatorId, defaultValue: "")
        createAt = values.decodeIfPresent(forKey: .createAt, defaultValue: 0)
        deleteAt = values.decodeIfPresent(forKey: .deleteAt, defaultValue: 0)
        displayName = values.decodeIfPresent(forKey: .displayName, defaultValue: "")
        extraUpdateAt = values.decodeIfPresent(forKey: .extraUpdateAt, defaultValue: 0)
        groupConstrained = values.decodeIfPresent(forKey: .groupConstrained, defaultValue: false)
        header = values.decodeIfPresent(forKey: .header, defaultValue: "")
        lastPostAt = values.decodeIfPresent(forKey: .lastPostAt, defaultValue: 0)
        lastRootPostAt = values.decodeIfPresent(forKey: .lastRootPostAt, defaultValue: 0)
        name = values.decodeIfPresent(forKey: .name, defaultValue: "")
        policyId = values.decodeIfPresent(forKey: .policyId, defaultValue: "")
        let propsData = try? values.decode([String:Any].self, forKey: .props)
        props = Database.default.json(from: propsData) ?? "{}"
        purpose = values.decodeIfPresent(forKey: .purpose, defaultValue: "")
        schemeId = values.decodeIfPresent(forKey: .schemeId, defaultValue: "")
        shared = values.decodeIfPresent(forKey: .shared, defaultValue: false)
        teamId = values.decodeIfPresent(forKey: .teamId, defaultValue: "")
        totalMsgCount = values.decodeIfPresent(forKey: .totalMsgCount, defaultValue: 0)
        totalMsgCountRoot = values.decodeIfPresent(forKey: .totalMsgCountRoot, defaultValue: 0)
        type = values.decodeIfPresent(forKey: .type, defaultValue: "O")
        updateAt = values.decodeIfPresent(forKey: .updateAt, defaultValue: 0)
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: ChannelKeys.self)
        try container.encode(self.id, forKey: .id)
        try container.encode(self.creatorId, forKey: .creatorId)
        try container.encode(self.createAt, forKey: .createAt)
        try container.encode(self.deleteAt, forKey: .deleteAt)
        try container.encode(self.displayName, forKey: .displayName)
        try container.encode(self.extraUpdateAt, forKey: .extraUpdateAt)
        try container.encode(self.groupConstrained, forKey: .groupConstrained)
        try container.encode(self.header, forKey: .header)
        try container.encode(self.lastPostAt, forKey: .lastPostAt)
        try container.encode(self.lastRootPostAt, forKey: .lastRootPostAt)
        try container.encode(self.name, forKey: .name)
        try container.encode(self.policyId, forKey: .policyId)
        try container.encode(self.props, forKey: .props)
        try container.encode(self.purpose, forKey: .purpose)
        try container.encode(self.schemeId, forKey: .schemeId)
        try container.encode(self.shared, forKey: .shared)
        try container.encode(self.teamId, forKey: .teamId)
        try container.encode(self.totalMsgCount, forKey: .totalMsgCount)
        try container.encode(self.totalMsgCountRoot, forKey: .totalMsgCountRoot)
        try container.encode(self.type, forKey: .type)
        try container.encode(self.updateAt, forKey: .updateAt)
    }
}
